File: vi/codes/swift/chapter_greedy/max_capacity.swift
/**
 * File: max_capacity.swift
 * Created Time: 2023-09-03
 * Author: nuomi1 (nuomi1@qq.com)
 */

/* Dung tích lớn nhất: Giải thuật tham lam */
func maxCapacity(ht: [Int]) -> Int {
    // Khởi tạo i, j nằm ở hai đầu của mảng
    var i = ht.startIndex, j = ht.endIndex - 1
    // Dung tích lớn nhất ban đầu là 0
    var res = 0
    // Lặp để lựa chọn tham lam cho đến khi hai tấm ván gặp nhau
    while i < j {
        // Cập nhật dung tích lớn nhất
        let cap = min(ht[i], ht[j]) * (j - i)
        res = max(res, cap)
        // Di chuyển tấm ván ngắn hơn vào trong
        if ht[i] < ht[j] {
            i += 1
        } else {
            j -= 1
        }
    }
    return res
}

@main
enum MaxCapacity {
    /* Mã điều khiển */
    static func main() {
        let ht = [3, 8, 5, 2, 7, 7, 3, 4]

        // Giải thuật tham lam
        let res = maxCapacity(ht: ht)
        print("Maximum capacity is \(res)")
    }
}

File: vi/codes/cpp/chapter_greedy/max_capacity.cpp
/**
 * File: max_capacity.cpp
 * Created Time: 2023-07-21
 * Author: krahets (krahets@163.com)
 */

#include "../utils/common.hpp"

/* Dung tích lớn nhất: Giải thuật tham lam */
int maxCapacity(vector<int> &ht) {
    // Khởi tạo i, j nằm ở hai đầu của mảng
    int i = 0, j = ht.size() - 1;
    // Dung tích lớn nhất ban đầu là 0
    int res = 0;
    // Lặp để chọn theo tham lam cho đến khi hai tấm ván gặp nhau
    while (i < j) {
        // Cập nhật dung tích lớn nhất
        int cap = min(ht[i], ht[j]) * (j - i);
        res = max(res, cap);
        // Di chuyển tấm ván ngắn hơn vào trong
        if (ht[i] < ht[j]) {
            i++;
        } else {
            j--;
        }
    }
    return res;
}

/* Driver Code */
int main() {
    vector<int> ht = {3, 8, 5, 2, 7, 7, 3, 4};

    // Giải thuật tham lam
    int res = maxCapacity(ht);
    cout << "Maximum capacity is " << res << endl;

    return 0;
}
